[CUMULUS-2453] Store errors for executions and granules and files in a JSON file

botanical opened the pull request:
**Summary:** Summary of changes

Addresses [CUMULUS-2453: Investigate data-migration2 performance testing execution migration failures](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2453)

## Changes

* Create `storeErrors` function which stores errors in system bucket
* Modify tf files to account for use of stackName and system bucket
* Packaged and deployed changes and this is an example of the output:
`data-migration2-executions-errors.json`
* Updated `granulesAndFiles.ts` and `executions.ts` to call `storeErrors`
```
{ "errors": [
"Could not create execution record in RDS for Dynamo execution arn arn:aws:states:us-east-1:[redacted]:execution:jtran-int-tf-DiscoverAndQueuePdrsExecutionPrefix:40fcf686-5e77-453f-bf7c-a4b0b4becc77:, Cause: RecordDoesNotExist: Record in collections with identifiers {\"name\":\"MOD09GQ_test-jtran-int-tf-IngestFromPdr-1614183240869\",\"version\":\"006\"} does not exist.",
"Could not create execution record in RDS for Dynamo execution arn arn:aws:states:us-east-1:[redacted]:execution:jtran-int-tf-DiscoverGranules:082dede3-5633-4225-b044-188de6314b20:, Cause: RecordDoesNotExist: Record in collections with identifiers {\"name\":\"collection-name-8daff3ef4a\",\"version\":\"collection-version-f074cc489b\"} does not exist."
]}
```
`data-migraiton2-granulesAndFiles-errors.json`
```
{ "errors": [
"Could not create granule record and file records in RDS for DynamoDB Granule granuleId: L2_HR_PIXC_product_0001-of-4154_test-jtran-int-tf-KinesisTestTrigger-1611879735452 with files [{\"bucket\":\"not-exist-bucket\",\"fileName\":\"L2_HR_PIXC_product_0001-of-4154.h5\",\"size\":50,\"checksum\":\"cd043418380acdf69b2fb80649357cda\",\"checksumType\":\"md5\",\"source\":\"s3://cumulus-test-sandbox-internal/somepath/L2_HR_PIXC_product_0001-of-4154.h5\",\"type\":\"data\"}], Cause: RecordDoesNotExist: Record in collections with identifiers {\"name\":\"L2_HR_PIXC_test-jtran-int-tf-KinesisTestTrigger-1611879735452\",\"version\":\"000\"} does not exist.",
"Could not create granule record and file records in RDS for DynamoDB Granule granuleId: L2_HR_PIXC_product_0001-of-4154_test-jtran-int-tf-KinesisTestTrigger-1611628271474 with files [{\"bucket\":\"not-exist-bucket\",\"fileName\":\"L2_HR_PIXC_product_0001-of-4154.h5\",\"size\":50,\"checksum\":\"cd043418380acdf69b2fb80649357cda\",\"checksumType\":\"md5\",\"source\":\"s3://cumulus-test-sandbox-internal/somepath/L2_HR_PIXC_product_0001-of-4154.h5\",\"type\":\"data\"}], Cause: RecordDoesNotExist: Record in collections with identifiers {\"name\":\"L2_HR_PIXC_test-jtran-int-tf-KinesisTestTrigger-1611628271474\",\"version\":\"000\"} does not exist."
]}
```

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Adhoc testing
- [ ] Integration tests

botanical commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,25 @@
+ * Store migration errors in JSON file in S3
+ * @param {string} bucket
+ * @param {string[]} message
+ * @param {string} recordClassification
```
Definitely open to better names because I am not sure this is it

Jkovarik replied:
I spent longer than I should trying to improve on that, and came up empty.   This seems fine, particularly in this case as it really doesn't present a readability concern.

markdboyd replied:
I would ask for descriptions of each param in the JSdoc though

markdboyd replied:
Also, `migrationName`?

botanical commented on lambdas/data-migration2/src/granulesAndFiles.ts:
```diff
@@ -139,16 +140,22 @@ export const migrateFileRecord = async (
-  loggingInterval: number
+  loggingInterval: number,
+  stackName?: string,
+  bucket?: string
```
I recognize that this function's parameter list is getting longer and I am wondering if I've now added too many parameters to it. Suggestions or comments welcome

Jkovarik replied:
Generally when they get > 3 or so arguments, it's worth considering making it a params object.   In this case, *particularly* as you have stackName/Bucket as optional values.

Jkovarik replied:
So something like 

```
const someFunc = async ( params: { 
  foo: string, 
  bar: string 
}) => { 
  const { foo, bar } = params;
......
```

Jkovarik commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,27 @@
+  message: string[],
+  recordClassification: string,
+  stackName?: string) => {
+  const file = `{
```
Consider using `putJsonS3Object` from the S3 client.

Jkovarik commented on lambdas/data-migration2/main.tf:
```diff
@@ -89,10 +89,12 @@ resource "aws_lambda_function" "data_migration2" {
+      GranulesTable               = var.dynamo_tables.granules.name
+      PdrsTable                   = var.dynamo_tables.pdrs.name
+      dbHeartBeat                 = var.rds_connection_heartbeat
+      system_bucket               = var.system_bucket
```
I'm wondering if we should use a more generic configuration variable name here as the output doesn't *have* to go to the system bucket, so much as *a* bucket.

Jkovarik commented on lambdas/data-migration2/src/executions.ts:
```diff
@@ -79,6 +82,8 @@ export const migrateExecutions = async (
     failed: 0,
     skipped: 0,
   };
+  const errorFile = [];
```
Nit: errorMessages or errors?    It's not really a 'file' or filestream object, so much as an array of errorMessages.

Jkovarik commented on lambdas/data-migration2/src/executions.ts:
```diff
@@ -97,13 +102,15 @@ export const migrateExecutions = async (
-          error
-        );
+        errorMessage = `Could not create execution record in RDS for Dynamo execution arn ${record.arn}:`;
+        errorFile.push(`Error: ${error} ${errorMessage}`);
```
Could this approach exceed the lambda/container's memory capacity for a reasonably large numbers of skips/errors (?) and if so I'm wondering if we should consider using a write stream or another defensive approach?

markdboyd replied:
+1. I think using a write stream would be safer.

Jkovarik reviewed (changes requested):
This looks really good!      I wanted to deploy/run and ran into some stack issues, so commenting with a few items for you to look at, will pick it back up in the AM re: deployment/testing.

markdboyd commented on lambdas/data-migration2/src/executions.ts:
```diff
@@ -97,13 +102,15 @@ export const migrateExecutions = async (
       }
     }
 
+    if (bucket) {
```
Is there a reason for this logic to be optional?

botanical replied:
Not any good reason

markdboyd commented on lambdas/data-migration2/src/executions.ts:
```diff
@@ -67,6 +68,8 @@ export const migrateExecutions = async (
   knex: Knex
 ): Promise<MigrationSummary> => {
   const executionsTable = envUtils.getRequiredEnvVar('ExecutionsTable', env);
+  const bucket = process.env.system_bucket;
+  const stackName = process.env.stackName;
```
I'm inclined to use `envUtils.getRequiredEnvVar` to fetch both of these values. They should always be set, since we're the ones setting them on the lambda deployment config, so treating them as required should be safe

markdboyd replied:
More fundamentally, because of the value of this error logging for users, I don't think it should be optional

botanical replied:
Agreed

markdboyd commented on lambdas/data-migration2/src/executions.ts:
```diff
@@ -79,6 +82,8 @@ export const migrateExecutions = async (
     skipped: 0,
   };
+  const errorFile = [];
+  let errorMessage;
```
Why are we initializing the errorMessage variable here? The variable seems to be only used inside the else block below, so I would prefer to use `const` inside that block

markdboyd commented on lambdas/data-migration2/tests/test-storeErrors.js:
```diff
@@ -0,0 +1,38 @@
+  const filename = `data-migration2-${recordClassification}-errors.json`;
+  const key = `${process.env.stackName}/${filename}`;
+
+  const item = await s3().getObject({
+    Bucket: process.env.system_bucket,
+    Key: key,
+  }).promise();
+  const messageBody = JSON.parse(item.Body);
```
Use the `getJsonS3Object` helper from the aws-client package

markdboyd commented on lambdas/data-migration2/tests/test-executions.js:
```diff
@@ -514,3 +515,38 @@ test.serial('migrateExecutions logs summary of migration for a specified logging
   t.true(logSpy.calledWith('Batch of 1 execution records processed, 1 total'));
 });
+
+test.serial('migrateExecutions calls storeError when system bucket is provided', async (t) => {
```
```suggestion
test.serial('migrateExecutions writes errors to S3 object', async (t) => {
```

markdboyd commented on lambdas/data-migration2/tests/test-granulesAndFiles.js:
```diff
@@ -590,6 +591,36 @@ test.serial('migrateGranulesAndFiles processes all non-failing granule records a
   t.is(fileRecords.length, 1);
 });
 
+test.serial('migrateGranulesAndFiles calls storeError when system bucket is provided', async (t) => {
```
```suggestion
test.serial('migrateGranulesAndFiles writes errors to an S3 object', async (t) => {
```

markdboyd commented on lambdas/data-migration2/tests/test-executions.js:
```diff
@@ -514,3 +515,38 @@ test.serial('migrateExecutions logs summary of migration for a specified logging
+  await migrateExecutions(process.env, t.context.knex);
+
+  // Check that error file exists in S3
+  const item = await s3().getObject({
+    Bucket: process.env.system_bucket,
+    Key: key,
+  }).promise();
+  const messageBody = JSON.parse(item.Body);
```
Use `getJsonS3Object` helper

markdboyd commented on lambdas/data-migration2/tests/test-executions.js:
```diff
@@ -514,3 +515,38 @@ test.serial('migrateExecutions logs summary of migration for a specified logging
+  await Promise.all([
+    // Have to use Dynamo client directly because creating
+    // via model won't allow creation of an invalid record
+    dynamodbDocClient().put({
```
I was curious and I took a quick look to find out if using `dynamodbDocClient().put` was really necessary. It is not. You can just use `executionsModel.create`. Furthermore, if the purpose of this test is just to ensure that an error is written for one failure, then we don't need the parent/child setup. Just creating an execution that tries to reference a non-existent async operation would be sufficient

markdboyd commented on lambdas/data-migration2/tests/test-executions.js:
```diff
@@ -514,3 +515,38 @@ test.serial('migrateExecutions logs summary of migration for a specified logging
+  }).promise();
+  const messageBody = JSON.parse(item.Body);
+
+  t.truthy(messageBody.errors[0]);
```
A truthy evaluation isn't telling us how many errors were written, which seems like a more relevant confirmation that things worked as expected? So maybe we should test the length of the written errors

markdboyd commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,27 @@
+  const file = `{
+    "errors": ${JSON.stringify(message)}
+  }`;
+  const filename = `data-migration2-${recordClassification}-errors.json`;
```
I think that the filename should be timestamped, so that each run of the migration has a separate error log that you can refer to.

markdboyd commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,27 @@
+  bucket: string,
+  message: string[],
+  recordClassification: string,
+  stackName?: string) => {
```
Why is `stackName` optional?

markdboyd commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,27 @@
+    "errors": ${JSON.stringify(message)}
+  }`;
+  const filename = `data-migration2-${recordClassification}-errors.json`;
+  const key = `${stackName}/${filename}` || filename;
```
What is the logic here doing? Effectively, even if `stackName` was undefined, you'd be doing:

```
`/${filename}` || filename
```

Whether filename has a value or not, nothing after the `||` is ever reached, so I think it's unnnecessary

markdboyd commented on lambdas/data-migration2/src/granulesAndFiles.ts:
```diff
@@ -172,10 +179,13 @@ export const migrateGranuleAndFilesViaTransaction = async (
+      logger.error(errorMessage, error);
+    }
+
+    if (bucket) {
```
Out of curiosity: is there any way once this file is written that the user will easily be able to find groups of "related" errors? Such as "X number of granules failed bc of this missing collection", etc. 

Is it easy enough to somehow parse that information from the generated report? 

I think it might be possible to generate a report that "classifies" the errors by cause by throwing/catching specific error types, but not sure if that's overkill

botanical commented on lambdas/data-migration2/README.md:
```diff
@@ -43,4 +43,6 @@ Migration summary:
\ No newline at end of file
+Records skipped indicates if a record was already migrated (which the lambda will throw with a `RecordAlreadyMigrated` error.)
+
+For `executions`, `granules`, and `files`, all data migration error messages will be written to JSON files `data-migration2-execution-errors.json` and `data-migration2-granulesAndFiles-errors.json` and uploaded to S3.
```
Ah, that's a good point! Will update

Jkovarik commented on lambdas/data-migration2/src/granulesAndFiles.ts:
```diff
@@ -135,20 +136,35 @@ export const migrateFileRecord = async (
   const files = dynamoRecord.files;
   const { granulesSummary, filesSummary } = granuleAndFileMigrationSummary;
+  const errorFile = [];
+  let errorMessage;
```
This doesn't appear to be used anywhere but in the catch below, consider moving it to around L182 where it's used rather than defining it here.

Jkovarik commented on lambdas/data-migration2/src/executions.ts:
```diff
@@ -67,6 +68,8 @@ export const migrateExecutions = async (
 ): Promise<MigrationSummary> => {
   const executionsTable = envUtils.getRequiredEnvVar('ExecutionsTable', env);
+  const bucket = envUtils.getRequiredEnvVar('system_bucket', env);
+  const stackName = envUtils.getRequiredEnvVar('stackName', env);
```
Side note/:ant: For the purposes of this PR this is fine given the nature of this code.  

I've been trying to avoid creating situations where we're implicitly relying on env variables deeper in the code for switchable configuration where we don't have to, as those are essentially >globals<.

We might consider putting this block in the main lambda function and then include these as a parameter/part of a params object in `migrateExecutions`, `migrateGranulesAndFiles`, and `migratePdrs` for example.  

Again, not *this* PR, but something to consider for maintenance reasons in the future/a future arch discussion.

Jkovarik reviewed (changes requested):
Just a couple of additional comments, nothing significant over the prior review.     Let me know if you have questions!

botanical commented:
@Jkovarik , @markdboyd I think I addressed most of the feedback. Doing adhoc testing has revealed that the generated JSON file for the granulesAndFiles errors are missing a `]}` at the end of the file so it cannot be parsed to JSON. I am not sure why it isn't being written because I definitely write it https://github.com/nasa/cumulus/pull/2179/files#diff-91b3c5306441fb2380946f4cabdfa533f0143379d13bebd235dc5cd9af8d8586R304 so another pair of eyes on that would be much appreciated.

markdboyd commented on lambdas/data-migration2/README.md:
```diff
@@ -43,4 +43,6 @@ Migration summary:
 
-Records skipped indicates if a record was already migrated (which the lambda will throw with a `RecordAlreadyMigrated` error.)
\ No newline at end of file
+Records skipped indicates if a record was already migrated (which the lambda will throw with a `RecordAlreadyMigrated` error.)
```
FWIW the above report summary structure is no longer accurate based on your previous refactors. To be honest I'm not sure if it's necessary to include a preview of the lambda output in the README

markdboyd replied:
In fact a lot of the README above is now inaccurate based on how this Lambda gets run. We don't have to clean it up in this PR necessarily, but need to make sure that we don't forget

botanical replied:
That's a good point. As for the lambda output, why do you think it's unnecessary to not include a preview of the output? I don't have strong views either way.

markdboyd replied:
Just because it seems like documentation that gets easily out of date and I'm not sure that it's the job of a README to give a preview of Lambda output. I see the README as describing mostly what something is and how to use it, if that is even applicable

botanical replied:
[`c6c1f32` (#2179)](https://github.com/nasa/cumulus/pull/2179/commits/c6c1f32eff6ceb5882a6dbb716014d89e99b8866#diff-4c92a92285abf37daad11f4a41b7ac4bf43e8649a39a04347600c035825cd2e9R20)

markdboyd commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,34 @@
+export const storeErrors = async (params: {
+  bucket: string,
+  filename: string,
+  recordClassification: string,
```
Nit: I'm gonna recommend `migrationName` as this variable name vs `recordClassification`, since I think that's a clearer explanation of what it is

markdboyd commented on lambdas/data-migration2/src/granulesAndFiles.ts:
```diff
@@ -266,14 +284,25 @@ export const migrateGranulesAndFiles = async (
+      errorFileWriteStream.write(',\n');
+    }
   }
+  errorFileWriteStream.write('\n]}');
```
```suggestion
  errorFileWriteStream.end('\n]}');
```

Based on https://nodejs.org/api/stream.html#stream_writable_end_chunk_encoding_callback

markdboyd commented on lambdas/data-migration2/README.md:
```diff
@@ -43,4 +43,6 @@ Migration summary:
\ No newline at end of file
+Records skipped indicates if a record was already migrated (which the lambda will throw with a `RecordAlreadyMigrated` error.)
+
+For `executions`, `granules`, and `files`, all data migration error messages will be written to JSON files `data-migration2-execution-errors.json` and `data-migration2-granulesAndFiles-errors.json` and uploaded to S3 to your configured system bucket at `${prefix}/data-migration-${type}-errors.json`.
```
These filenames would have timestamps in them as well

markdboyd replied:
```suggestion
For `executions`, `granules`, and `files`, all data migration error messages will be written to JSON files `data-migration2-execution-errors.json` and `data-migration2-granulesAndFiles-errors.json` and uploaded to S3 to your configured system bucket at `${prefix}/data-migration-${type}-errors-${timestamp}.json`.
```

markdboyd commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,34 @@
+import moment from 'moment';
```
We're using moment but I don't see that we've added it as a dependency to the package.json for the Lambda?

markdboyd commented on lambdas/data-migration2/package.json:
```diff
@@ -11,7 +11,7 @@
     "build-lambda-zip": "cd dist/webpack && node ../../../../bin/zip.js lambda.zip index.js",
     "package": "npm run clean && npm run prepare && npm run webpack && npm run build-lambda-zip",
-    "test": "../../node_modules/.bin/ava",
+    "test": "../../node_modules/.bin/ava --serial",
```
I would prefer not to set this here in the package config and instead just add `.serial` to every test where we need it

botanical replied:
Hm, I am willing to remove this but doing so doesn't achieve the same thing.

markdboyd replied:
Fair enough, but if we need it then we have some unexpected interference among our tests and I'd like to understand why

markdboyd commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,34 @@
+ * @param {string} params.migrationName - Name of migration
+ * @param {string} params.stackName - User stack name/prefix
+ * @param {string | undefined} params.timestamp - Timestamp for unit testing
+ * @returns {void}
```
```suggestion
 * @returns {Promise<void>}
```

botanical replied:
https://github.com/nasa/cumulus/pull/2179/files#diff-56c5b1137921652643573bffebcd7fae07a919d6517811b49ab0f84759ba10beR44

markdboyd commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,34 @@
+ *
+ * @param {Object} params
+ * @param {string} params.bucket - Name of S3 bucket where file will be uploaded
+ * @param {string[]} params.filepath - Write Stream file path
```
```suggestion
 * @param {string} params.filepath - Write Stream file path
```

botanical replied:
https://github.com/nasa/cumulus/pull/2179/files#diff-56c5b1137921652643573bffebcd7fae07a919d6517811b49ab0f84759ba10beR40

markdboyd commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,34 @@
+  const { bucket, filepath, migrationName, stackName, timestamp } = params;
+  const fileKey = `data-migration2-${migrationName}-errors`;
+  const dateString = timestamp || moment.utc().format('YYYY-MM-DD_HH:MM:SS');
+  const key = `${stackName}/${fileKey}_${dateString}.json`;
```
Nit: let's be consistent in our use of dashes vs underscores

```suggestion
  const key = `${stackName}/${fileKey}-${dateString}.json`;
```

botanical replied:
https://github.com/nasa/cumulus/pull/2179/files#diff-56c5b1137921652643573bffebcd7fae07a919d6517811b49ab0f84759ba10beR56

markdboyd commented on lambdas/data-migration2/src/granulesAndFiles.ts:
```diff
@@ -230,6 +245,9 @@ export const migrateGranulesAndFiles = async (
     granulesResult: granuleMigrationResult,
     filesResult: fileMigrationResult,
   };
+  const filepath = 'granulesAndFilesMigrationErrorLog.json';
+  const errorFileWriteStream = fs.createWriteStream(filepath);
+  errorFileWriteStream.write('{ "errors": [\n');
```
I'm on the fence, but I might abstract this logic into a `createErrorFileWriteStream` helper in the `storeErrors.ts` file. Just make it accept `migrationName` as an argument.

botanical replied:
https://github.com/nasa/cumulus/pull/2179/files#diff-56c5b1137921652643573bffebcd7fae07a919d6517811b49ab0f84759ba10beR15

markdboyd commented on lambdas/data-migration2/src/granulesAndFiles.ts:
```diff
@@ -266,14 +284,25 @@ export const migrateGranulesAndFiles = async (
+      errorFileWriteStream.write(',\n');
+    }
   }
+  errorFileWriteStream.end('\n]}');
```
Likewise, this could be a `closeErrorFileWriteStream` helper

botanical replied:
Do you mind expanding on why a helper to end the write stream is useful? It would only be one line, right? I am

botanical replied:
I ended up adding a end function to help ensure that the stream closes based on some feedback I got from @Jkovarik [`963d9f7` (#2179)](https://github.com/nasa/cumulus/pull/2179/commits/963d9f7adde0399981f6a5ecdb6b5ee44eca2086)

markdboyd commented on CHANGELOG.md:
```diff
@@ -152,10 +152,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
     - Updated `cumulus-tf/variables.tf` to include `optional_dynamo_tables` that will be merged with `dynamo_tables`.
+  - **CUMULUS-2453**
+    - Created `storeErrors` function which stores errors in system bucket.
+    - Updated `executions` and `granulesAndFiles` to call `storeErrors` to store migration errors.
```
```suggestion
    - Updated `executions` and `granulesAndFiles` data migrations to call `storeErrors` to store migration errors.
```

botanical replied:
https://github.com/nasa/cumulus/pull/2179/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR157

markdboyd commented on CHANGELOG.md:
```diff
@@ -152,10 +152,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
     - Added summary type file `packages/db/src/types/summary.ts` with `MigrationSummary` and `DataMigration1` and `DataMigration2` types.
     - Updated `data-migration1` and `data-migration2` lambdas to return `MigrationSummary` objects.
     - Added logging for every batch of 100 records processed for executions, granules and files, and PDRs.
+    - Removed `RecordAlreadyMigrated` logs in `data-migration1` and `data-migration2`
```
This line seems to duplicate the one below it?

botanical replied:
[`c6c1f32` (#2179)](https://github.com/nasa/cumulus/pull/2179/commits/c6c1f32eff6ceb5882a6dbb716014d89e99b8866#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edL164)

markdboyd commented on lambdas/data-migration2/src/executions.ts:
```diff
@@ -97,16 +111,21 @@ export const migrateExecutions = async (
-          error
-        );
+        const errorMessage = `Could not create execution record in RDS for Dynamo execution arn ${record.arn}:`;
+        errorFileWriteStream.write(JSON.stringify(`Error: ${error} ${errorMessage}`));
```
```suggestion
        errorFileWriteStream.write(JSON.stringify(`${errorMessage}, Cause: ${error}`));
```

botanical replied:
https://github.com/nasa/cumulus/pull/2179/commits/7ea4144651c1375087efef89938fe560d9b30d29

markdboyd commented on lambdas/data-migration2/src/granulesAndFiles.ts:
```diff
@@ -230,6 +245,9 @@ export const migrateGranulesAndFiles = async (
     granulesResult: granuleMigrationResult,
     filesResult: fileMigrationResult,
   };
+  const filepath = 'granulesAndFilesMigrationErrorLog.json';
```
This filename (here and in the executions migration) should probably be timestamped. The race condition you're seeing is probably due to this filename (which is the same throughout all tests) is getting unlinked while it's still being written to by some tests

botanical replied:
The race condition didn't go away with timestamped names but it occurred significantly less. What I ended up implementing was a function to close the write stream and ensure it finishes [`963d9f7` (#2179)](https://github.com/nasa/cumulus/pull/2179/commits/963d9f7adde0399981f6a5ecdb6b5ee44eca2086#diff-56c5b1137921652643573bffebcd7fae07a919d6517811b49ab0f84759ba10beR29)

botanical commented on lambdas/data-migration2/src/granulesAndFiles.ts:
```diff
@@ -175,12 +186,12 @@ export const migrateGranuleAndFilesViaTransaction = async (
-        error
-      );
+
+      stream.write(JSON.stringify(`Error: ${error} ${errorMessage}`));
```
https://github.com/nasa/cumulus/pull/2179/commits/7ea4144651c1375087efef89938fe560d9b30d29

markdboyd commented on lambdas/data-migration2/README.md:
```diff
@@ -17,30 +17,8 @@ The resources migrated in Data Migration 2 are:
+
+Records skipped indicates if a record was already migrated (which the lambda will throw with a `RecordAlreadyMigrated` error.)
+
+For `executions`, `granules`, and `files`, all data migration error messages will be written to JSON files `data-migration2-execution-errors-${timestamp}.json` and `data-migration2-granulesAndFiles-errors-${timestamp}.json` and uploaded to S3 to your configured system bucket at `${prefix}/data-migration-${type}-errors-${timestamp}.json`.
```
```suggestion
For `executions`, `granules`, and `files`, all data migration error messages will be written to JSON files `data-migration2-execution-errors-${timestamp}.json` and `data-migration2-granulesAndFiles-errors-${timestamp}.json` and uploaded to S3 to your configured system bucket at `${prefix}/data-migration2-${type}-errors-${timestamp}.json`.
```

markdboyd commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,63 @@
+}) => {
+  const { bucket, filepath, migrationName, stackName, timestamp } = params;
+  const fileKey = `data-migration2-${migrationName}-errors`;
+  const dateString = timestamp || moment.utc().format('YYYY-MM-DD_HH:MM:SS');
```
Can we avoid using moment and introducing the dependency by doing `new Date().toISOString()`?

markdboyd replied:
It also might be nice for this function to log the name of the object it wrote to S3 for traceability

markdboyd reviewed (approved):
Had a couple more nits but I'll approve pending their resolution

Jkovarik reviewed (changes requested):
This all looks good, but on verification, I'm seeing output like: 

```
 nsidc-jkovarik  :45 ->cat data-migration2-granulesAndFiles-errors-2021-04-27T13\:41\:31.225Z.json  |more
{ "errors": [
,
,
,
,
,
,
"Could not create granule record and file records in RDS for DynamoDB Granule granuleId: MOD09GQ.A0370686.tq5B1P.006.4526292427657 with files 
<snip> 
```

which is resulting in: 

 nsidc-jkovarik  :46 ->cat data-migration2-granulesAndFiles-errors-2021-04-27T13\:41\:31.225Z.json  |jq
parse error: Expected value before ',' at line 2, column 1

I think we need to add a skip for entries without errors/filter the results that are resulting in an empty error value.

botanical commented:
> This all looks good, but on verification, I'm seeing output like:
> 
> ```
>  nsidc-jkovarik  :45 ->cat data-migration2-granulesAndFiles-errors-2021-04-27T13\:41\:31.225Z.json  |more
> { "errors": [
> ,
> ,
> ,
> ,
> ,
> ,
> "Could not create granule record and file records in RDS for DynamoDB Granule granuleId: MOD09GQ.A0370686.tq5B1P.006.4526292427657 with files 
> <snip> 
> ```
> 
> which is resulting in:
> 
> nsidc-jkovarik :46 ->cat data-migration2-granulesAndFiles-errors-2021-04-27T13:41:31.225Z.json |jq
> parse error: Expected value before ',' at line 2, column 1
> 
> I think we need to add a skip for entries without errors/filter the results that are resulting in an empty error value.

Thanks for the feedback @Jkovarik! That was a necessary callout. I added some tests for skipped records and failed records and changed the implementation to use a package `JSONStream`'s [stringify](https://www.npmjs.com/package/JSONStream#jsonstreamstringifyopen-sep-close ) function which handles the delimiting and stores the errors in an array.

markdboyd commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,71 @@
+  const dateString = timestamp || new Date().toISOString();
+  const filepath = `${migrationName}ErrorLog-${dateString}.json`;
+  const errorFileWriteStream = fs.createWriteStream(filepath);
+  const jsonWriteStream = JSONStream.stringify();
+  jsonWriteStream.pipe(errorFileWriteStream);
+  errorFileWriteStream.write('{ "errors": \n');
```
This is a little confusing to me. We're writing to `errorFileWriteStream` directly, even though it seems like what we want to do is write to the JSONstream, which is piped to the error file already.

I think we could avoid this if we customized the open/close options that we pass when initializing the JSONStream, like so:

```suggestion
  const jsonWriteStream = JSONStream.stringify('{"errors": [\n', '\n,\n', '\n]}\n');
  jsonWriteStream.pipe(errorFileWriteStream);
  errorFileWriteStream.write('{ "errors": \n');
```

If we do this, then `closeErrorFileWriteStream` should also be unnecessary, if I'm understanding the docs for JSONstream correctly, because the whole point of passing a custom "close" string when setting up the JSON stream is letting it handle things for you automatically

botanical replied:
Thanks for the feedback! I am skeptical that I can let `stringify` end the stream, so I am still making an explicit call to end the JSON stream but I removed the closeErrorFileWriteStream function. [`3e09741` (#2179)](https://github.com/nasa/cumulus/pull/2179/commits/3e09741a7febe8ee0442fa9d57d6ff5a9d2a8910)

markdboyd commented on lambdas/data-migration2/src/granulesAndFiles.ts:
```diff
@@ -175,12 +188,12 @@ export const migrateGranuleAndFilesViaTransaction = async (
-        error
-      );
+
+      stream.write(JSON.stringify(`${errorMessage}, Cause: ${error}`));
```
I think the JSON.stringify here is unnecessary. Part of what the JSONStream is doing for you is this stringification automatically.

With the JSON.stringify in place, when I inspect the local contents of the error file, I see:

```
{
  "errors": [
    "\"Could not create granule record and file records in RDS for DynamoDB Granule granuleId: 9a2c4 with files [{\\\"bucket\\\":\\\"ba67b2bd6a\\\",\\\"fileName\\\":\\\"MOD09GQ.A4369670.7bAGCH.006.0739896140643.hdf\\\",\\\"size\\\":1098034,\\\"checksum\\\":\\\"checkSum01\\\",\\\"checksumType\\\":\\\"md5\\\",\\\"source\\\":\\\"source\\\",\\\"type\\\":\\\"data\\\",\\\"key\\\":\\\"fc75692596\\\"}], Cause: RecordDoesNotExist: Record in collections with identifiers {\\\"name\\\":\\\"6fe57\\\",\\\"version\\\":\\\"0.0.0\\\"} does not exist.\"",
    "\"Could not create granule record and file records in RDS for DynamoDB Granule granuleId: f0095 with files [{\\\"bucket\\\":\\\"ba67b2bd6a\\\",\\\"fileName\\\":\\\"MOD09GQ.A4369670.7bAGCH.006.0739896140643.hdf\\\",\\\"size\\\":1098034,\\\"checksum\\\":\\\"checkSum01\\\",\\\"checksumType\\\":\\\"md5\\\",\\\"source\\\":\\\"source\\\",\\\"type\\\":\\\"data\\\",\\\"key\\\":\\\"851a0de80f\\\"}], Cause: RecordDoesNotExist: Record in collections with identifiers {\\\"name\\\":\\\"dd859\\\",\\\"version\\\":\\\"0.0.0\\\"} does not exist.\""
  ]
}
```

You can see that it is unnecessarily double-escaping the contents of each error message, which is less readable.

With the JSON.stringify removed, I see:

```
{
  "errors": [
    "Could not create granule record and file records in RDS for DynamoDB Granule granuleId: a7c71 with files [{\"bucket\":\"e824412d07\",\"fileName\":\"MOD09GQ.A4369670.7bAGCH.006.0739896140643.hdf\",\"size\":1098034,\"checksum\":\"checkSum01\",\"checksumType\":\"md5\",\"source\":\"source\",\"type\":\"data\",\"key\":\"f7d6ebe20e\"}], Cause: RecordDoesNotExist: Record in collections with identifiers {\"name\":\"a07de\",\"version\":\"0.0.0\"} does not exist.",
    "Could not create granule record and file records in RDS for DynamoDB Granule granuleId: 24935 with files [{\"bucket\":\"e824412d07\",\"fileName\":\"MOD09GQ.A4369670.7bAGCH.006.0739896140643.hdf\",\"size\":1098034,\"checksum\":\"checkSum01\",\"checksumType\":\"md5\",\"source\":\"source\",\"type\":\"data\",\"key\":\"90520d5818\"}], Cause: RecordDoesNotExist: Record in collections with identifiers {\"name\":\"6207d\",\"version\":\"0.0.0\"} does not exist."
  ]
}
```

Which is I think more what we want

markdboyd commented on lambdas/data-migration2/src/granulesAndFiles.ts:
```diff
@@ -140,20 +143,30 @@ export const migrateFileRecord = async (
-): Promise<GranulesAndFilesMigrationResult> => {
-  const files = dynamoRecord.files ?? [];
+  loggingInterval: number,
+  stream: any,
```
I think we can give this a type of `Writable` from the `stream` package. Also, we should give this variable a more descriptive name, like `errorsLogWriteStream`

markdboyd commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,58 @@
+  const jsonWriteStream = JSONStream.stringify('{"errors": [\n', '\n,', '\n]}\n');
+  jsonWriteStream.pipe(errorFileWriteStream);
+
+  return { jsonWriteStream, errorFileWriteStream, filepath };
```
We don't need to include `errorFileWriteStream` in the return here

Jkovarik replied:
@markdboyd   Are you suggesting the explicit close/await on the finished stream isn't required?    Apologies if that's obtuse, I'm not sure I'm following though.

markdboyd commented on lambdas/data-migration2/src/storeErrors.ts:
```diff
@@ -0,0 +1,58 @@
+    Body: fs.createReadStream(filepath),
+  }).promise();
+
+  logger.info(`Stored error log file with key ${key} to bucket ${bucket}.`);
```
```suggestion
  logger.info(`Stored error log file on S3 at s3://${bucket}/${key}`);
```

This would make it easier for someone to copy/paste the S3 location into their terminal and use it with a CLI command

botanical replied:
Ooh that is smart!

markdboyd commented on lambdas/data-migration2/src/granulesAndFiles.ts:
```diff
@@ -266,14 +292,21 @@ export const migrateGranulesAndFiles = async (
     await searchQueue.shift();
     record = await searchQueue.peek();
   }
+  jsonWriteStream.end();
```
It seems like we might need to bring back the close stream helper. I'm getting intermittent unit test failures locally without it which I think are related. Sorry for asking you to remove that.

We just don't need to pass any arguments to the `stream.end()` call.

Jkovarik reviewed (approved):
Thanks for addressing my concerns, this looks good 👍🏻
